
# World Demographics
## [Link to the Dataset](https://www.kaggle.com/datasets/kushagraarya10/cia-global-statistical-database?select=demographics_data.csv)

The dataset specifically focuses on the demographics of different countries. It includes categories such as population growth, infant mortality, sex ratio, and etc. The information in the dataset can be used to have an inside look into the bigger picture for different countries.

### Importing the Data

In [12]:
import pandas as pd
import numpy as np
df = pd.read_csv("demographics_data.csv")
print("\n Read the csv.file:\n",df)


 Read the csv.file:
             Country Total_Population Population_Growth_Rate  Birth_Rate  \
0       AFGHANISTAN         39232003                  2.26%       34.79   
1          AKROTIRI              NaN                    NaN         NaN   
2           ALBANIA          3101621                  0.19%       12.48   
3           ALGERIA         44758398                  1.27%       17.84   
4    AMERICAN SAMOA            44620                  1.74%       16.18   
..              ...              ...                    ...         ...   
254       WEST BANK          3176549                   2.1%       28.30   
255           WORLD       7979261010                  1.03%       18.10   
256           YEMEN         31565602                  1.83%       24.10   
257          ZAMBIA         20216029                  2.86%       34.50   
258        ZIMBABWE         15418674                  1.99%       29.40   

     Death_Rate  Net_Migration_Rate  Median_Age  Sex_Ratio  \
0         12.08

### Basic Data Exploration

In [2]:
print('These are the first 5 rows:')
df.head()

These are the first 5 rows:


Country Total_Population Population_Growth_Rate  Birth_Rate  \
0     AFGHANISTAN         39232003                  2.26%       34.79   
1        AKROTIRI              NaN                    NaN         NaN   
2         ALBANIA          3101621                  0.19%       12.48   
3         ALGERIA         44758398                  1.27%       17.84   
4  AMERICAN SAMOA            44620                  1.74%       16.18   

   Death_Rate  Net_Migration_Rate  Median_Age  Sex_Ratio  \
0       12.08                0.10        19.5       1.02   
1         NaN                 NaN         NaN        NaN   
2        7.36                3.22        34.3       0.97   
3        4.33                0.81        28.9       1.03   
4        6.19               27.36        27.2       0.99   

   Infant_Mortality_Rate  Total_Fertility_Rate Total_Literacy_Rate  \
0                 103.06                  4.53               37.3%   
1                    NaN                   NaN                 NaN   
2                  10.54                  1.55               98.4%   
3                  19.22                  2.47               81.4%   
4                   9.87                  2.13                 NaN   

  Male_Literacy_Rate Female_Literacy_Rate Youth_Unemployment_Rate  
0              39.4%                 7.2%                   20.2%  
1                NaN                  NaN                     NaN  
2              38.8%                   6%                   27.8%  
3              41.3%                 0.7%                   31.9%  
4                NaN                  NaN                     NaN

In [3]:
print('These are the last 5 rows:')
df.tail()

These are the last 5 rows:


Country Total_Population Population_Growth_Rate  Birth_Rate  \
254  WEST BANK          3176549                   2.1%        28.3   
255      WORLD       7979261010                  1.03%        18.1   
256      YEMEN         31565602                  1.83%        24.1   
257     ZAMBIA         20216029                  2.86%        34.5   
258   ZIMBABWE         15418674                  1.99%        29.4   

     Death_Rate  Net_Migration_Rate  Median_Age  Sex_Ratio  \
254         3.5                 3.9        21.7       1.03   
255         7.7                 NaN        31.0       1.01   
256         5.5                 0.2        21.6       1.02   
257         6.0                 0.2        18.2       1.00   
258         6.7                 2.9        21.0       0.95   

     Infant_Mortality_Rate  Total_Fertility_Rate Total_Literacy_Rate  \
254                   16.1                  3.54               97.5%   
255                   30.8                  2.42               86.7%   
256                   45.5                  2.91               70.1%   
257                   36.3                  4.49               86.7%   
258                   34.0                  3.51               89.7%   

    Male_Literacy_Rate Female_Literacy_Rate Youth_Unemployment_Rate  
254              98.8%                96.2%                   39.6%  
255              90.1%                83.3%                   17.9%  
256              32.5%                 8.1%                   25.5%  
257              25.1%                 3.7%                   26.1%  
258              21.8%                 1.5%                    7.3%

In [4]:
print("The format of reading how many rows and columns there are is this: (rows,columns)")
df.shape

The format of reading how many rows and columns there are is this: (rows,columns)


(259, 14)

In [5]:
print("The column names are:")
print(df.columns)

The column names are:
Index(['Country', 'Total_Population', 'Population_Growth_Rate', 'Birth_Rate',
       'Death_Rate', 'Net_Migration_Rate', 'Median_Age', 'Sex_Ratio',
       'Infant_Mortality_Rate', 'Total_Fertility_Rate', 'Total_Literacy_Rate',
       'Male_Literacy_Rate', 'Female_Literacy_Rate',
       'Youth_Unemployment_Rate'],
      dtype='object')


### A Closer Look at the Data

In [6]:
print("Here is some General Information")
df.info()

Here is some General Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Country                  259 non-null    object 
 1   Total_Population         237 non-null    object 
 2   Population_Growth_Rate   237 non-null    object 
 3   Birth_Rate               228 non-null    float64
 4   Death_Rate               230 non-null    float64
 5   Net_Migration_Rate       229 non-null    float64
 6   Median_Age               227 non-null    float64
 7   Sex_Ratio                227 non-null    float64
 8   Infant_Mortality_Rate    227 non-null    float64
 9   Total_Fertility_Rate     227 non-null    float64
 10  Total_Literacy_Rate      168 non-null    object 
 11  Male_Literacy_Rate       212 non-null    object 
 12  Female_Literacy_Rate     212 non-null    object 
 13  Youth_Unemployment_Rate  206 non-null    object

The columns that receive float64 under the dtype section shows that those columns use numbers that contain decimals which falls under quantitative data. The columns that receive object under the dtype section means that there is not a numerical type which would lead to it being classified as categorical data. The data type in columns Total_Population, Population_Growth_Rate, Total_Literacy_Rate, Male_Literacy_Rate, Female_Literacy_Rate, and Youth_Unemployment_Rate should be converted into numerical data because most of these are percentages that ended up showing as object under data type because it had the percentage symbol written with it. Total_Population should be converted into a quantitative data type because it is filled with integers. 

In [7]:
print("Here are the Summary Statistics:")
df.describe()

Here are the Summary Statistics:


Birth_Rate   Death_Rate  Net_Migration_Rate  Median_Age   Sex_Ratio  \
count   228.000000   230.000000          229.000000  227.000000  227.000000   
mean     26.361842    25.018565            3.657511   31.833480    1.004229   
std     132.909844   187.367894            5.565715    9.236467    0.187592   
min       5.800000     1.420000           -2.400000   14.800000    0.830000   
25%      10.800000     5.702500            0.630000   24.050000    0.950000   
50%      15.030000     7.255000            1.890000   31.700000    0.980000   
75%      22.225000     9.100000            4.400000   39.500000    1.010000   
max    2020.000000  2021.000000           45.800000   55.400000    3.340000   

       Infant_Mortality_Rate  Total_Fertility_Rate  
count             227.000000            227.000000  
mean               19.367665             11.292555  
std                18.431299            133.983614  
min                 1.500000              1.090000  
25%                 5.755000              1.680000  
50%                12.000000              1.970000  
75%                28.525000              2.800000  
max               103.060000           2021.000000

There are some things that are unusual or inconsistent which are the max for Birth_Rate and Death_Rate. The Birth rate definition is the number of births per year out of every 1,000 people. The max is 2,020 for Birth Rate which would mean that one country delievered 2,020 births out of 1,000 people. That seems highly unlikely and suspicious. If you look at the min, 25%, 50%, and 75% they seem to be in reason with each other. They slowly build off of the other. The max is an extreme jump away from what was shown prior. The max death rate is 2,021 which would mean that for one country out of 1,000 people 2,021 died. That is a very high number and unlikely. If you look at the min, 25%, 50%, and 75% they seem to be in reason with each other. They slowly build off of the other. The max is an extreme jump away from what was shown prior. The same pattern which can be seen in Birth and Death Rate can also be seen in the max for Net_Migration_Rate.

### Missing and Duplicate Data

In [8]:
print("This is the number of missing data for each column.")
missing_data = df.isnull().sum()
print(missing_data)


This is the number of missing data for each column.
Country                     0
Total_Population           22
Population_Growth_Rate     22
Birth_Rate                 31
Death_Rate                 29
Net_Migration_Rate         30
Median_Age                 32
Sex_Ratio                  32
Infant_Mortality_Rate      32
Total_Fertility_Rate       32
Total_Literacy_Rate        91
Male_Literacy_Rate         47
Female_Literacy_Rate       47
Youth_Unemployment_Rate    53
dtype: int64


In [9]:
percentage_missing_data = df.isnull().sum()/len(df)*100 
print("This is the percentage of missing data for each column")
print(percentage_missing_data)

This is the percentage of missing data for each column
Country                     0.000000
Total_Population            8.494208
Population_Growth_Rate      8.494208
Birth_Rate                 11.969112
Death_Rate                 11.196911
Net_Migration_Rate         11.583012
Median_Age                 12.355212
Sex_Ratio                  12.355212
Infant_Mortality_Rate      12.355212
Total_Fertility_Rate       12.355212
Total_Literacy_Rate        35.135135
Male_Literacy_Rate         18.146718
Female_Literacy_Rate       18.146718
Youth_Unemployment_Rate    20.463320
dtype: float64


In [10]:
duplicated_data = df.duplicated().sum()
print(f'Number of duplicates:{duplicated_data}')

Number of duplicates:0


### Adding a New Column

In [43]:
df_new = df.copy()

# Converting Total_Population from object to numbers 
if df_new['Total_Population'].dtype == 'object':
    df_new['Total_Population'] = df_new['Total_Population'].str.replace(',', '').astype(float)

# Removing the % from Population_Growth_Rate and turning it into a float
if df_new['Population_Growth_Rate'].dtype == 'object':
    df_new['Population_Growth_Rate'] = df_new['Population_Growth_Rate'].str.replace('%', '').astype(float) / 100
else:
    # This section makes sure that it is a percentage by being less than 1
    if df_new['Population_Growth_Rate'].any() > 1:
        df_new['Population_Growth_Rate'] = df_new['Population_Growth_Rate'] / 100

df_new['Predicted_Population_Next_Year'] = df_new['Total_Population'] * (1 + df_new['Population_Growth_Rate'])
print(df_new)
df_new.head()

            Country  Total_Population  Population_Growth_Rate  Birth_Rate  \
0       AFGHANISTAN      3.923200e+07                  0.0226       34.79   
1          AKROTIRI               NaN                     NaN         NaN   
2           ALBANIA      3.101621e+06                  0.0019       12.48   
3           ALGERIA      4.475840e+07                  0.0127       17.84   
4    AMERICAN SAMOA      4.462000e+04                  0.0174       16.18   
..              ...               ...                     ...         ...   
254       WEST BANK      3.176549e+06                  0.0210       28.30   
255           WORLD      7.979261e+09                  0.0103       18.10   
256           YEMEN      3.156560e+07                  0.0183       24.10   
257          ZAMBIA      2.021603e+07                  0.0286       34.50   
258        ZIMBABWE      1.541867e+07                  0.0199       29.40   

     Death_Rate  Net_Migration_Rate  Median_Age  Sex_Ratio  \
0         12.

Country  Total_Population  Population_Growth_Rate  Birth_Rate  \
0     AFGHANISTAN        39232003.0                  0.0226       34.79   
1        AKROTIRI               NaN                     NaN         NaN   
2         ALBANIA         3101621.0                  0.0019       12.48   
3         ALGERIA        44758398.0                  0.0127       17.84   
4  AMERICAN SAMOA           44620.0                  0.0174       16.18   

   Death_Rate  Net_Migration_Rate  Median_Age  Sex_Ratio  \
0       12.08                0.10        19.5       1.02   
1         NaN                 NaN         NaN        NaN   
2        7.36                3.22        34.3       0.97   
3        4.33                0.81        28.9       1.03   
4        6.19               27.36        27.2       0.99   

   Infant_Mortality_Rate  Total_Fertility_Rate Total_Literacy_Rate  \
0                 103.06                  4.53               37.3%   
1                    NaN                   NaN                 NaN   
2                  10.54                  1.55               98.4%   
3                  19.22                  2.47               81.4%   
4                   9.87                  2.13                 NaN   

  Male_Literacy_Rate Female_Literacy_Rate Youth_Unemployment_Rate  \
0              39.4%                 7.2%                   20.2%   
1                NaN                  NaN                     NaN   
2              38.8%                   6%                   27.8%   
3              41.3%                 0.7%                   31.9%   
4                NaN                  NaN                     NaN   

   Predicted_Population_Next_Year  
0                    4.011865e+07  
1                             NaN  
2                    3.107514e+06  
3                    4.532683e+07  
4                    4.539639e+04

To create a new column. I turned both the total population and population growth rates into floats and removed symbols such as "," and "%" that made them strings. The new column is the predicted population of next year by adding 1 to the population growth rate and then multiplying that by the total population. An example would be total population= 6, growth rate=0.03, (1+0.03)=1.03 * 6= 6.18. The predicted population for the next year based on those numbers would be 6.18. Something I can explore further is how to take it out of the equation and just display the real number. 

### Filtering Data

In [58]:
df_filtered = df_new[df_new['Net_Migration_Rate'] > 20.00]
df_filtered = df_filtered.sort_values(by='Net_Migration_Rate', ascending=False)
print("The countries with the highest Net_Migration_Rate:")
print(df_filtered[['Country','Net_Migration_Rate']].head())

The countries with the highest Net_Migration_Rate:
                             Country  Net_Migration_Rate
225                            SYRIA               45.80
241                          UKRAINE               37.30
4                     AMERICAN SAMOA               27.36
57                      COOK ISLANDS               26.20
151  MICRONESIA, FEDERATED STATES OF               20.92


In [69]:
df_filtered2=df_new[df_new['Death_Rate']>15.00]
df_filtered2 = df_filtered2.sort_values(by='Death_Rate', ascending=False)
print("The countries with the highest Death Rate:")
print(df_filtered2[['Country','Death_Rate']].head())

The countries with the highest Death Rate:
            Country  Death_Rate
78   EUROPEAN UNION     2021.00
232         TOKELAU     2021.00
241         UKRAINE       19.80
138       LITHUANIA       15.17
204          SERBIA       15.10


I created a subset which looked at the countries with the highest Migration Rate and another one which looked at the countries with the highest Death Rate. I looked at these two sections because I was curious if there was a link between migration and death. Specifically, if the migration rate was high because the death rate was high. The only country that appears in both subsets are Ukraine. It is notable that European Union and Tokelau are extremely higher than the rest for the Death Rate subset. For future inspection I should remove them to get a clearer list.

In [74]:
print(df_new['Country'].unique())

['AFGHANISTAN' 'AKROTIRI' 'ALBANIA' 'ALGERIA' 'AMERICAN SAMOA' 'ANDORRA'
 'ANGOLA' 'ANGUILLA' 'ANTARCTICA' 'ANTIGUA AND BARBUDA' 'ARCTIC OCEAN'
 'ARGENTINA' 'ARMENIA' 'ARUBA' 'ASHMORE AND CARTIER ISLANDS'
 'ATLANTIC OCEAN' 'AUSTRALIA' 'AUSTRIA' 'AZERBAIJAN' 'BAHAMAS, THE'
 'BAHRAIN' 'BANGLADESH' 'BARBADOS' 'BELARUS' 'BELGIUM' 'BELIZE' 'BENIN'
 'BERMUDA' 'BHUTAN' 'BOLIVIA' 'BOSNIA AND HERZEGOVINA' 'BOTSWANA'
 'BOUVET ISLAND' 'BRAZIL' 'BRITISH INDIAN OCEAN TERRITORY'
 'BRITISH VIRGIN ISLANDS' 'BRUNEI' 'BULGARIA' 'BURKINA FASO' 'BURMA'
 'BURUNDI' 'CABO VERDE' 'CAMBODIA' 'CAMEROON' 'CANADA' 'CAYMAN ISLANDS'
 'CENTRAL AFRICAN REPUBLIC' 'CHAD' 'CHILE' 'CHINA' 'CHRISTMAS ISLAND'
 'CLIPPERTON ISLAND' 'COCOS (KEELING) ISLANDS' 'COLOMBIA' 'COMOROS'
 'CONGO, DEMOCRATIC REPUBLIC OF THE' 'CONGO, REPUBLIC OF THE'
 'COOK ISLANDS' 'CORAL SEA ISLANDS' 'COSTA RICA' 'CROATIA' 'CUBA'
 'CURACAO' 'CYPRUS' 'CZECHIA' 'DENMARK' 'DHEKELIA' 'DJIBOUTI' 'DOMINICA'
 'DOMINICAN REPUBLIC' 'ECUADOR' 'EGYPT' 'EL SALVAD

Although 7 of the columns are listed as categorical. Only really the "Country" column is filled with categorical variables. The other columns that are listed as objects under data type have the percentage symbol which classified them as categorical except "Total_Population", it does not have any percentage and is a column full of integers. I'm unsure why it is classified as an object but this can be changed. Knowing the different countries is still very important for grouping because it gives so many different pairings for countries and geography. 

### Observations/Final Thoughts

The world demographics dataset overall is pretty interesting. I'm particularly drawn to the Death_Rate and Net_Migration_Rate. It makes me curious as to the why behind these percentages. For example I know Ukraine was high for both the Net_Migration_Rate and the Death_Rate columns. The data was taken from 2024-2025. It can be assummed that both are high due to the Russia-Ukraine War that is still going on. I think I could also compare Death_Rate to the Median_Age because the median age does to some extent reflect how long people in a country live to. Notable issues I ran into were the % symbols being in the data, the death rate for the European Union and Tokelau, and some missing data for certain countries. The percentage symbol classified a lot of data as object or categorical. If the data were written with a decimal point instead of % it would have been easier to process and subtract a few steps. The European Union and Tokelau are both 2021 for death rate which is incredibly to high compared to the other data. I believe it is an error. Lastly, some variables were missing in different columns. When I further dive into this project I can either delete them or come up with different ways to work around it. But in general this dataset works out great, my group chose well. 